File: c++-tcp/log.cc
#include <glog/logging.h>

#include "log.h"

using InstanceState::COMMITTED;
using InstanceState::EXECUTED;

bool IsCommitted(Instance const& instance) {
  return instance.state_ == InstanceState::COMMITTED;
}
bool IsExecuted(Instance const& instance) {
  return instance.state_ == InstanceState::EXECUTED;
}
bool IsInProgress(Instance const& instance) {
  return instance.state_ == InstanceState::INPROGRESS;
}

bool operator==(Command const& a, Command const& b) {
  return a.type_ == b.type_ && a.key_ == b.key_ && a.value_ == b.value_;
}

bool operator==(Instance const& a, Instance const& b) {
  return a.ballot_ == b.ballot_ && a.index_ == b.index_ &&
         a.client_id_ == b.client_id_ && a.state_ == b.state_ &&
         a.command_ == b.command_;
}

bool Insert(std::unordered_map<int64_t, Instance>* log, Instance instance) {
  auto i = instance.index_;
  auto it = log->find(i);
  if (it == log->end()) {
    (*log)[i] = std::move(instance);
    return true;
  }
  if (IsCommitted(it->second) || IsExecuted(it->second)) {
    CHECK(it->second.command_ == instance.command_) << "Insert case2";
    return false;
  }
  if (instance.ballot_ > it->second.ballot_) {
    (*log)[i] = std::move(instance);
    return false;
  }
  if (instance.ballot_ == it->second.ballot_)
    CHECK(it->second.command_ == instance.command_) << "Insert case3";
  return false;
}

void Log::Append(Instance instance) {
  std::scoped_lock lock(mu_);

  int64_t i = instance.index_;
  if (i <= global_last_executed_)
    return;

  if (Insert(&log_, std::move(instance))) {
    last_index_ = std::max(last_index_, i);
    cv_committable_.notify_all();
  }
}

void Log::Commit(int64_t index) {
  CHECK(index > 0) << "invalid index";

  std::unique_lock lock(mu_);
  auto it = log_.find(index);
  while (it == log_.end()) {
    cv_committable_.wait(lock);
    it = log_.find(index);
  }

  if (IsInProgress(it->second))
    it->second.state_ = COMMITTED;

  if (IsExecutable())
    cv_executable_.notify_one();
}

std::optional<std::tuple<int64_t, kvstore::KVResult>> Log::Execute() {
  std::unique_lock lock(mu_);
  while (running_ && !IsExecutable())
    cv_executable_.wait(lock);

  if (!running_)
    return std::nullopt;

  auto it = log_.find(last_executed_ + 1);
  CHECK(it != log_.end());
  Instance* instance = &it->second;
  kvstore::KVResult result =
      kvstore::Execute(instance->command_, kv_store_.get());
  instance->state_ = EXECUTED;
  ++last_executed_;
  return {{instance->client_id_, std::move(result)}};
}

void Log::CommitUntil(int64_t leader_last_executed, int64_t ballot) {
  CHECK(leader_last_executed >= 0) << "invalid leader_last_executed";
  CHECK(ballot >= 0) << "invalid ballot";

  std::unique_lock lock(mu_);
  for (auto i = last_executed_ + 1; i <= leader_last_executed; ++i) {
    auto it = log_.find(i);
    if (it == log_.end())
      break;
    CHECK(ballot >= it->second.ballot_) << "CommitUntil case 2";
    if (it->second.ballot_ == ballot)
      it->second.state_ = COMMITTED;
  }
  if (IsExecutable())
    cv_executable_.notify_one();
}

void Log::TrimUntil(int64_t leader_global_last_executed) {
  std::scoped_lock lock(mu_);
  while (global_last_executed_ < leader_global_last_executed) {
    ++global_last_executed_;
    auto it = log_.find(global_last_executed_);
    CHECK(it != log_.end() && IsExecuted(it->second)) << "TrimUntil case 1";
    log_.erase(it);
  }
}

std::vector<Instance> Log::Instances() const {
  std::scoped_lock lock(mu_);
  std::vector<Instance> instances;
  for (auto i = global_last_executed_ + 1; i <= last_index_; ++i)
    if (auto it = log_.find(i); it != log_.end())
      instances.push_back(it->second);
  return instances;
}

Instance const* Log::at(std::size_t i) const {
  auto it = log_.find(i);
  return it == log_.end() ? nullptr : &it->second;
}

std::unordered_map<int64_t, Instance> Log::GetLog() {
  std::unordered_map<int64_t, Instance> local_log(log_);
  return local_log;
}

File: go-tcp/log/log.go
package log

import (
	"github.com/psu-csl/replicated-store/go/kvstore"
	tcp "github.com/psu-csl/replicated-store/go/multipaxos/network"
	logger "github.com/sirupsen/logrus"
	"sync"
)

func IsCommitted(instance *tcp.Instance) bool {
	return instance.State == tcp.Committed
}

func IsExecuted(instance *tcp.Instance) bool {
	return instance.State == tcp.Executed
}

func IsInProgress(instance *tcp.Instance) bool {
	return instance.State == tcp.Inprogress
}

func IsEqualCommand(cmd1, cmd2 *tcp.Command) bool {
	return cmd1.Type == cmd2.Type && cmd1.Key == cmd2.Key &&
		cmd1.Value == cmd2.Value
}

func IsEqualInstance(a, b *tcp.Instance) bool {
	return a.Ballot == b.Ballot && a.Index == b.Index &&
		a.ClientId == b.ClientId && a.State == b.State &&
		IsEqualCommand(a.Command, b.Command)
}

func Insert(log map[int64]*tcp.Instance, instance *tcp.Instance) bool {
	i := instance.Index
	if _, ok := log[i]; !ok {
		log[i] = instance
		return true
	}

	if IsCommitted(log[i]) || IsExecuted(log[i]) {
		if !IsEqualCommand(log[i].Command, instance.Command) {
			logger.Panicf("case 2 violation\n")
		}
		return false
	}

	if instance.Ballot > log[i].Ballot {
		log[i] = instance
		return false
	}

	if instance.Ballot == log[i].Ballot {
		if !IsEqualCommand(log[i].Command, instance.Command) {
			logger.Panicf("case 3 violation\n")
		}
	}
	return false
}

type Log struct {
	running            bool
	kvStore            kvstore.KVStore
	log                map[int64]*tcp.Instance
	lastIndex          int64
	lastExecuted       int64
	globalLastExecuted int64
	mu                 sync.Mutex
	cvExecutable       *sync.Cond
	cvCommittable      *sync.Cond
}

func NewLog(s kvstore.KVStore) *Log {
	l := Log{
		running:            true,
		kvStore:            s,
		log:                make(map[int64]*tcp.Instance),
		lastIndex:          0,
		lastExecuted:       0,
		globalLastExecuted: 0,
		mu:                 sync.Mutex{},
	}
	l.cvExecutable = sync.NewCond(&l.mu)
	l.cvCommittable = sync.NewCond(&l.mu)
	return &l
}

func (l *Log) LastExecuted() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	return l.lastExecuted
}

func (l *Log) GlobalLastExecuted() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	return l.globalLastExecuted
}

func (l *Log) AdvanceLastIndex() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	l.lastIndex += 1
	return l.lastIndex
}

func (l *Log) SetLastIndex(lastIndex int64) {
	l.mu.Lock()
	defer l.mu.Unlock()

	if lastIndex > l.lastIndex {
		l.lastIndex = lastIndex
	}
}

func (l *Log) LastIndex() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.lastIndex
}

func (l *Log) Stop() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.running = false
	l.kvStore.Close()
	l.cvExecutable.Signal()
}

func (l *Log) IsExecutable() bool {
	instance, ok := l.log[l.lastExecuted+1]
	if ok && IsCommitted(instance) {
		return true
	}
	return false
}

func (l *Log) Append(instance *tcp.Instance) {
	l.mu.Lock()
	defer l.mu.Unlock()

	i := instance.Index
	if i <= l.globalLastExecuted {
		return
	}

	if Insert(l.log, instance) {
		if i > l.lastIndex {
			l.lastIndex = i
		}
		l.cvCommittable.Broadcast()
	}
}

func (l *Log) Commit(index int64) {
	if index <= 0 {
		logger.Panicf("Index %v < 0\n", index)
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	instance, ok := l.log[index]
	for !ok {
		l.cvCommittable.Wait()
		instance, ok = l.log[index]
	}

	if IsInProgress(instance) {
		instance.State = tcp.Committed
	}
	if l.IsExecutable() {
		l.cvExecutable.Signal()
	}
}

func (l *Log) ReadInstance() *tcp.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()

	for l.running && !l.IsExecutable() {
		l.cvExecutable.Wait()
	}

	if !l.running {
		return nil
	}

	instance, ok := l.log[l.lastExecuted+1]
	if !ok {
		logger.Panicf("Instance at Index %v empty\n", l.lastExecuted+1)
	}
	instance.State = tcp.Executed
	l.lastExecuted += 1
	return instance
}

func (l *Log) Execute(instance *tcp.Instance) (int64, *kvstore.KVResult) {
	result := kvstore.Execute(instance.Command, l.kvStore)
	return instance.ClientId, &result
}

func (l *Log) CommitUntil(leaderLastExecuted int64, ballot int64) {
	if leaderLastExecuted < 0 {
		logger.Panic("invalid leader_last_executed in commit_until")
	}
	if ballot < 0 {
		logger.Panic("invalid ballot in commit_until")
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	for i := l.lastExecuted + 1; i <= leaderLastExecuted; i++ {
		instance, ok := l.log[i]
		if !ok {
			break
		}
		if ballot < instance.Ballot {
			panic("CommitUntil case 2")
		}
		if instance.Ballot == ballot {
			instance.State = tcp.Committed
		}
	}
	if l.IsExecutable() {
		l.cvExecutable.Signal()
	}
}

func (l *Log) TrimUntil(leaderGlobalLastExecuted int64) {
	l.mu.Lock()
	defer l.mu.Unlock()

	for l.globalLastExecuted < leaderGlobalLastExecuted {
		l.globalLastExecuted += 1
		instance, ok := l.log[l.globalLastExecuted]
		if !ok || !IsExecuted(instance) {
			logger.Panicln("TrimUntil case 1")
		}
		delete(l.log, l.globalLastExecuted)
	}
}

func (l *Log) Instances() []*tcp.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()

	instances := make([]*tcp.Instance, 0, len(l.log))
	for i := l.globalLastExecuted + 1; i <= l.lastIndex; i++ {
		if i, ok := l.log[i]; ok {
			instance := *i
			instances = append(instances, &instance)
		}
	}
	return instances
}

func (l *Log) At(index int64) *tcp.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()
	if instance, ok := l.log[index]; ok {
		return instance
	}
	return nil
}

func (l *Log) GetLog() map[int64]*tcp.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()

	logMap := make(map[int64]*tcp.Instance)
	for index, instance := range l.log {
		copyInstance := *instance
		logMap[index] = &copyInstance
	}
	return logMap
}
